local/sdk-sriov/pkg/utils/utils.go

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package utils contains useful helper methods for network machinery
+package utils
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+
+	"github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+)
+
+var (
+	sysBusPci = "/sys/bus/pci/devices"
+)
+
+const (
+	totalVfFile      = "sriov_totalvfs"
+	configuredVfFile = "sriov_numvfs"
+)
+
+// IsSriovCapable check if a pci device SR-IOV capable given its pci address
+func IsSriovCapable(pciAddr string) bool {
+	totalVfFilePath := filepath.Join(sysBusPci, pciAddr, totalVfFile)
+	if _, err := os.Stat(totalVfFilePath); err != nil {
+		return false
+	}
+	// sriov_totalvfs file exists -> sriov capable
+	return true
+}
+
+// IsSriovVF check if a pci device has link to a PF
+func IsSriovVF(pciAddr string) bool {
+	totalVfFilePath := filepath.Join(sysBusPci, pciAddr, "physfn")
+	if _, err := os.Stat(totalVfFilePath); err != nil {
+		return false
+	}
+	return true
+}
+
+// GetVFconfigured returns number of VF configured for a PF
+func GetVFconfigured(pf string) int {
+	configuredVfPath := filepath.Join(sysBusPci, pf, configuredVfFile)
+	vfs, err := ioutil.ReadFile(filepath.Clean(configuredVfPath))
+	if err != nil {
+		return 0
+	}
+	configuredVFs := bytes.TrimSpace(vfs)
+	numConfiguredVFs, err := strconv.Atoi(string(configuredVFs))
+	if err != nil {
+		return 0
+	}
+	return numConfiguredVFs
+}
+
+// IsSriovConfigured returns true if sriov_numvfs reads > 0 else false
+func IsSriovConfigured(addr string) bool {
+	return GetVFconfigured(addr) > 0
+}
+
+// GetSriovVFcapacity returns SRIOV VF capacity - number of VFs that can be created for specified PF
+func GetSriovVFcapacity(pfPciAddr string) int {
+	totalVfFilePath := filepath.Join(sysBusPci, pfPciAddr, totalVfFile)
+	vfs, err := ioutil.ReadFile(filepath.Clean(totalVfFilePath))
+	if err != nil {
+		return 0
+	}
+	totalvfs := bytes.TrimSpace(vfs)
+	numvfs, err := strconv.Atoi(string(totalvfs))
+	if err != nil {
+		return 0
+	}
+	return numvfs
+}
+
+// DeviceExists validates PciAddr given as string and checks if it exists
+func DeviceExists(pciAddr string) error {
+	// Check system pci address
+
+	// sysbus pci address regex
+	var validLongID = regexp.MustCompile(`^0{4}:[0-9a-f]{2}:[0-9a-f]{2}.[0-7]{1}$`)
```

Comment:
let's use common local values and something like sync.Once to reduce regexp complication here.